pkg/compactor/blocks_cleaner_test.go

```diff
@@ -1266,104 +1267,180 @@ func TestBlocksCleaner_RaceCondition_CleanerUpdatesBucketIndexWhileAnotherCleane
-		Bucket: bucketClient,
-		preIterHook: func(_ context.Context, dir string, _ func(string) error, _ ...objstore.IterOption) {
-			// When listing blocks, wait until the 2nd cleaner started to list markers.
-			if path.Base(dir) == tenantID {
-				<-cleaner2ListMarkersStarted
-			}
+				cleaner2BucketClient = &hookBucket{
+					Bucket: bucketClient,
+					preIterHook: func() func(context.Context, string, func(string) error, ...objstore.IterOption) {
+						once := sync.Once{}
+
+						return func(_ context.Context, dir string, _ func(string) error, _ ...objstore.IterOption) {
+							if path.Base(dir) == block.MarkersPathname {
+								// When listing markers, signal that the 2nd cleaner started to list markers and
+								// then wait until it should proceed.
+								once.Do(func() {
+									close(cleaner2ListMarkersStarted)
+									<-cleaner2ListMarkersUnblocked
+								})
+							}
+						}
+					}(),
+				}
+
+				return
+			},
 		},
-		preDeleteHook: func() func(context.Context, string) {
-			once := sync.Once{}
-
-			return func(_ context.Context, name string) {
-				// When deleting the deletion mark of a block (which is expected to be the last object deleted of a block)
-				// unblock the 2nd cleaner markers listing with some delay. The delay is used to make sure this deletion
-				// completes before the 2nd cleaner start listing markers.
-				if path.Base(name) == block.DeletionMarkFilename {
-					once.Do(func() {
-						go func() {
-							time.Sleep(100 * time.Millisecond)
-							close(cleaner2ListMarkersUnblocked)
-						}()
-					})
+		"the 2nd cleaner fetches new deletion marks after the 1st cleaner has started deleting blocks for the new deletion marks": {
+			mockBucketClients: func(bucketClient objstore.Bucket) (cleaner1BucketClient, cleaner2BucketClient objstore.Bucket) {
+				cleaner2ListMarkersStarted := make(chan struct{})
+				cleaner2GetDeletionMarkUnblocked := make(chan struct{})
+
+				// Mock the 1st cleaner to delete deletion marks when the 2nd cleaner is between the listing
+				// of deletion marks and fetching their content (for new markers).
+				cleaner1BucketClient = &hookBucket{
+					Bucket: bucketClient,
+					preDeleteHook: func() func(context.Context, string) {
+						once := sync.Once{}
+
+						return func(_ context.Context, name string) {
+							if path.Base(name) == block.DeletionMarkFilename {
+								// When deleting the deletion mark of a block, wait until cleaner2 has listed the deletion marks,
+								// because we want cleaner2 to discover the deletion mark before we delete it.
+								<-cleaner2ListMarkersStarted
+
+								// Then signal the deletion happened with some delay. The delay is used to make
+								// sure this deletion completes before the 2nd cleaner is unblocked.
+								once.Do(func() {
+									go func() {
+										time.Sleep(100 * time.Millisecond)
+										close(cleaner2GetDeletionMarkUnblocked)
+									}()
+								})
```

Comment:
this looks more appropriate for a `postDeleteHook`. It will be less flaky

Reply:
Agree! I don't know why I haven't thought about it before. Done in https://github.com/grafana/mimir/pull/8905/commits/55b097c390f393f939793a330d2d5fd843cfd0b3

Reply:
I've re-run the test on top of `main` and both test cases fail.